langkit/dsl.py

```diff
@@ -517,21 +493,178 @@ def process_subclass(mcs, name, bases, dct, is_root):
         dct['_base'] = base
         dct['_env_spec'] = env_spec
         dct['_is_token_node'] = is_token_node
+        dct['_is_enum_node'] = is_enum_node
 
         # Make sure subclasses don't inherit the "list_type" cache from their
         # base classes.
         dct['_list_type'] = None
         dct['_element_type'] = element_type
         dct['_annotations'] = annotations
 
+        cls = type.__new__(mcs, name, bases, dct)
+
+        mcs.astnode_types.append(cls)
+
+        # Create the corresponding ASTNodeType subclass
+        if cls._base is _ASTNodeList:
+            # Only root list types are supposed to directly subclass
+            # _ASTNodeList.
+            element_type = cls._element_type._resolve()
+            assert element_type
+            astnode_type = element_type.list
+        else:
+            astnode_type = ASTNodeType(
+                cls._name, cls._location, cls._doc,
+                base=None if is_root else cls._base._resolve(),
+                fields=cls._fields,
+                env_spec=cls._env_spec,
+                annotations=cls._annotations,
+
+                # only enum nodes are abstract at this point
+                is_abstract=cls._is_enum_node,
+
+                is_enum_node=cls._is_enum_node,
+                is_bool_node=cls._is_enum_node and cls._qualifier,
+                is_token_node=cls._is_token_node
+            )
+
+        astnode_type.dsl_decl = cls
+        cls._type = astnode_type
+
+        if is_enum_node:
+            mcs.create_enum_node_alternatives(cls, astnode_type)
+
+        return cls
+
+    @classmethod
+    def import_enum_node_attributes(mcs, dct, qualifier, alts, fields):
+        from langkit.expressions import AbstractProperty
+        from langkit.parsers import _Row, _Transform, Opt
+
+        def create_parser_bool_node(cls, *args):
+            # If the node is a boolean node, then we want to parse the
+            # sub-parsers as an optional parser that will be booleanized.
+            return Opt(*args).as_bool(cls)
+
+        def create_parser_enum_node(cls, alt_typeref, *args):
+            # Otherwise, we want to parse the sub-parsers as a row + transform
+            return _Transform(_Row(*args), alt_typeref)
+
+        def constructor(cls, *args):
+            """
+            This constructor can be used in the grammar to create a parser for
+            this enum node. This is valid only when qualifier is set to True.
+            """
+            assert qualifier
+            return create_parser_bool_node(cls, *args)
+
+        dct['__new__'] = constructor
+        dct['_create_parser'] = classmethod(
+            create_parser_bool_node if qualifier
+            else create_parser_enum_node
+        )
+
+        dct['_alternatives'] = alts
+        dct['_qualifier'] = qualifier
+
+        # Make Alternative instances available as attributes of the enum node
+        # class for a convenient way to create parsers for them.
```

Comment:
Thinking out loud: at this point I wonder if there’s much value keeping these instances around. Maybe we should just put parser constructors here (as functions, not classes with __new__ methods) and be done with it. Anyway, that’s for the future :)